cloud/services/bastionhosts/bastionhosts.go

```diff
@@ -0,0 +1,178 @@
+	bastionHostSpec, ok := spec.(*Spec)
+	if !ok {
+		return errors.New("invalid Bastion Host specification")
+	}
+
+	klog.V(2).Infof("getting subnet %s", bastionHostSpec.SubnetName)
+	subnet, err := s.SubnetsClient.Get(ctx, s.Scope.ResourceGroup(), bastionHostSpec.VnetName, bastionHostSpec.SubnetName)
+	if err != nil {
+		return errors.Wrap(err, "failed to get subnet")
+	}
+
+	klog.V(2).Infof("successfully got subnet %s", bastionHostSpec.SubnetName)
+
+	klog.V(2).Infof("checking if public ip %s exist otherwise will try to create", bastionHostSpec.PublicIPName)
+	publicIP := network.PublicIPAddress{}
+	publicIP, err = s.PublicIPsClient.Get(ctx, s.Scope.ResourceGroup(), bastionHostSpec.PublicIPName)
+	if err != nil && azure.ResourceNotFound(err) {
+		iperr := s.createBastionPublicIP(ctx, bastionHostSpec.PublicIPName)
+		if iperr != nil {
+			return errors.Wrap(iperr, "failed to create bastion public IP")
+		}
+		var errPublicIP error
+		publicIP, errPublicIP = s.PublicIPsClient.Get(ctx, s.Scope.ResourceGroup(), bastionHostSpec.PublicIPName)
+		if errPublicIP != nil {
+			return errors.Wrap(errPublicIP, "failed to get public IP")
+		}
+	} else if err != nil {
+		return errors.Wrap(err, "failed to look for existing public IP")
+	}
+	klog.V(2).Infof("successfully got public ip %s", bastionHostSpec.PublicIPName)
+
+	bastionHostName := bastionHostSpec.Name
+	klog.V(2).Infof("creating bastion host %s", bastionHostName)
+
+	bastionHostIPConfigName := fmt.Sprintf("%s-%s", bastionHostSpec.Name, "bastionIP")
+
+	err = s.Client.CreateOrUpdate(
+		ctx,
+		s.Scope.ResourceGroup(),
+		bastionHostName,
+		network.BastionHost{
+			Name:     to.StringPtr(bastionHostName),
+			Location: to.StringPtr(s.Scope.Location()),
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.Name(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(bastionHostName),
+				Role:        to.StringPtr("Bastion"),
+			})),
+			BastionHostPropertiesFormat: &network.BastionHostPropertiesFormat{
+				DNSName: to.StringPtr(fmt.Sprintf("%s-bastion", strings.ToLower(bastionHostName))),
+				IPConfigurations: &[]network.BastionHostIPConfiguration{
+					{
+						Name: to.StringPtr(bastionHostIPConfigName),
+						BastionHostIPConfigurationPropertiesFormat: &network.BastionHostIPConfigurationPropertiesFormat{
+							Subnet: &network.SubResource{
+								ID: subnet.ID,
+							},
+							PublicIPAddress: &network.SubResource{
+								ID: publicIP.ID,
+							},
+							PrivateIPAllocationMethod: network.Static,
+						},
+					},
+				},
+			},
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, "cannot create bastion host")
+	}
+
+	klog.V(2).Infof("successfully created bastion host %s", bastionHostName)
+	return nil
+}
+
+// Delete deletes the bastion host with the provided scope.
+func (s *Service) Delete(ctx context.Context, spec interface{}) error {
```

Comment:
`Delete` will only delete the bastion service. If a public IP address was created in the `Reconcile` func, then it seems like it should be cleaned up here too.

Here is a similar scenario where a public IP is provided by the user. We should decided on a way to represent ownership of the Azure resource to help us to understand how to act during provisioning and deprovisioning.

/cc @alexeldeib and @CecileRobertMichon

Reply:
> Yes, I was thinking if the user needs to provide the IP or CAPZ will create it and I'm not sure the path to follow, that's is why I implemented in that way, if we decide the user need to provide the IP then I can remove the creation part.
I was following a similar approach that the VMs do as well to create the IP.
I don't have strong opinions on that, i think that should follow what the team wants for this.

I agree the IP should be created if it doesn't exist in a similar manor as VMs. Where this implementation differs is that it does not delete the IP when the bastion host is deleted. That means, the IP which is created in the `Create` func may not get deleted, but should be deleted if it was created in the `Create` func.

If created in `Create`, the IP should be deleted in `Delete` (imo).

Reply:
FYI #716 might impact this change... We might want to use the same Reconcile/Delete for Bastion IPs as other public IPs (see how Reconcile() is used for VM public IPs in #716) and use tags (similar to resource group and vnet tags) to mark the IP as "owned"/managed by CAPZ. When deleting IPs, it should only delete the ones that are owned by that CAPZ cluster.

Reply:
#716 merged so you should be able to reuse it now.

Reply:
@CecileRobertMichon updated this PR to use the new way for services